t/lib/OpenQA/Test/Utils.pm

```diff
@@ -50,6 +53,28 @@ our (@EXPORT, @EXPORT_OK);
     qw(run_cmd test_cmd)
 );
 
+# The function OpenQA::Utils::service_port method hardcodes ports in a
+# sequential range starting with OPENQA_BASE_PORT. This can cause problems
+# especially in repeated testing if any of the port in that range is already
+# occupied, so we inject random, free ports for the services here
+#
+# Potential point for
+# later improvement: In Mojo::IOLoop::Server::generate_port keep the sock
+# object on the port and reuse it in listen to prevent race condition
+#
+# Potentially this approach can also be used in production code.
+sub mock_service_ports {
+    my %ports;
+    Test::MockModule->new('OpenQA::Utils')->redefine(
+        service_port => sub {
+            my ($service) = @_;
+            my $port = $ports{$service} //= Mojo::IOLoop::Server->generate_port;
+            note("Mocking service port for $service to be $port");
+            return $port;
+        });
+    note('Used ports: ' . dumper(\%ports));
```

Comment:
just one question:  I would like to know, after calling `mock_service_ports`, and then call `service_port 'webui'`,  will the `%ports` will be set to `{webui => }` ?

Reply:
Yes, %ports is a hash for each service as identifier and a free port. The original function in OpenQA::Utils returns fixed ports instead which are used for production instances.